keras/src/callbacks/orbax_checkpoint.py

```diff
@@ -0,0 +1,702 @@
+                  metrics=['accuracy'])
+
+    EPOCHS = 10
+    checkpoint_dir = '/tmp/ckpt'
+    orbax_checkpoint_callback = keras.callbacks.OrbaxCheckpoint(
+        directory=checkpoint_dir,
+        monitor='val_accuracy',
+        mode='max',
+        save_best_only=True)
+
+    # Model is saved at the end of every epoch, if it's the best seen so far.
+    model.fit(epochs=EPOCHS, callbacks=[orbax_checkpoint_callback])
+
+    # The model can be loaded from a specific checkpoint step as -
+    checkpoint = keras.callbacks.OrbaxCheckpoint(directory=checkpoint_dir)
+    checkpoint.load_checkpoint(step=5, model=model)  # Load from step 5
+
+    # Alternatively, save checkpoints every N batches -
+    orbax_checkpoint_callback = keras.callbacks.OrbaxCheckpoint(
+        directory=checkpoint_dir,
+        save_freq=100)  # Save every 100 batches
+
+    model.fit(epochs=EPOCHS, callbacks=[orbax_checkpoint_callback])
+
+    # Or use a SaveDecisionPolicy for more control -
+    from orbax.checkpoint import checkpoint_managers
+    policy = checkpoint_managers.FixedIntervalPolicy(interval=5)
+    orbax_checkpoint_callback = keras.callbacks.OrbaxCheckpoint(
+        directory=checkpoint_dir,
+        save_decision_policy=policy)  # Save every 5 epochs
+
+    model.fit(epochs=EPOCHS, callbacks=[orbax_checkpoint_callback])
+    ```
+
+    Args:
+        directory: string, path to the directory where to save the checkpoints.
+        monitor: The metric name to monitor (e.g., 'val_loss').
+        verbose: Verbosity mode, 0 or 1.
+        save_best_only: if `save_best_only=True`, it only saves when the model
+            is considered the "best" based on the monitored quantity.
+        mode: one of {'auto', 'min', 'max'}. Used with `save_best_only`.
+        save_freq: `'epoch'` or integer. Frequency to save checkpoints.
+        max_to_keep: Integer, maximum number of recent checkpoints to keep.
+            If None, keeps all. Defaults to 5.
+        keep_period: Integer, keep one checkpoint every `keep_period` saves.
+            Useful for keeping checkpoints less frequently over long runs.
+        initial_value_threshold: Floating point initial "best" value for the
+            monitor, used with `save_best_only`.
+        save_optimizer_state: Boolean, whether to include optimizer variables
+            in the checkpoint. Defaults to True.
+        save_on_background: Boolean, whether to save asynchronously in the
+            background. Defaults to True.
+        save_metadata: Dict or callable, additional metadata to save with each
+            checkpoint. If callable, it will be called with (epoch, logs) and
+            should return a dict. Defaults to None.
+        save_data_iterator: Dict or callable, data iterator state to save with
+            each checkpoint. If callable, it will be called with (epoch, logs)
+            and should return a dict with serializable iterator state.
+            Defaults to None.
```

Comment:
Curious what this feature is for? It seems very similar to the one above.

Reply:
The save_data_iterator parameter is indeed similar to save_metadata, but serves a very specific and important purpose in training resumption scenarios.

Reply:
Can you give an example? We didn't have that in the regular checkpoint format, so I'm curious.

Reply:
I have added a test for this test_save_data_iterator_state, test_load_checkpoint_with_iterator_state , test_tensorflow_iterator_restoration &     def test_jax_iterator_restoration(self):